autocertkit/common.py

```diff
@@ -0,0 +1,158 @@
+    res = {"returncode": process.returncode, "stdout": stdout.strip(),
+           "stderr": str(stderr).strip()}
+    if logging:
+        log.debug("returncode: %d" % process.returncode)
+        log.debug("stdout: %s" % str(stdout))
+        log.debug("stderr: %s" % str(stderr))
+        if res["returncode"] != 0:
+            log.error(f"ERR: Could not make local call: {call}")
+
+    return res
+
+
+class SecureChannel:
+    """Wrap of ssh and scp"""
+
+    def __init__(self, ip, user, password, timeout=300):
+        self.ip = ip
+        self.user = user
+        self.password = password
+        self.timeout = timeout
+
+    def _wrap_cmd(self, cmd):
+        escape_cmd = cmd.replace('$', '\$')
+        return fr'''{EXPECT} << EOF
+set timeout {self.timeout}
+spawn {escape_cmd}
+expect {{
+    "continue connecting (yes/no)?" {{send "yes\n"; exp_continue}}
+    "password:" {{send "{self.password}\n"; exp_continue}}
+    eof {{catch wait result; exit [lindex \$result 3]}}
+    timeout {{exit 250}}
+}}
+EOF
+'''
+
+    def _wrap_ssh(self, cmd):
+        return self._wrap_cmd(f'{SSH} {self.user}@{self.ip} {{ {cmd} }}')
+
+    def run_cmd(self, cmd):
+        """Run command simply and ignore stderr"""
+        return make_local_call(self._wrap_ssh(cmd), shell=True, timeout=self.timeout)
+
+    def run_cmd_ext(self, cmd):
+        """Run command and capture stdout and stderr separately"""
+        id = str(uuid.uuid4())
+        fcmd = f'.ack_cmd.{id}'
+        frc = f'.ack_rc.{id}'
+        fout = f'.ack_out.{id}'
+        ferr = f'.ack_err.{id}'
+
+        with open(f'/tmp/{fcmd}', 'w') as f:
+            f.write(cmd)
+        self.put_file(f'/tmp/{fcmd}')
+
+        self.run_cmd(fr'sh {fcmd} >{fout} 2>{ferr}; echo "$?" >{frc}')
+
+        self.get_file(f'.ack_*.{id}', '/tmp/')
+        contents = []
+        for f in [frc, fout, ferr]:
+            with open(f'/tmp/{f}', 'r') as f:
+                contents.append(f.read().strip())
+
+        self.run_cmd(fr'rm -f .ack_*.{id}')
+        for f in [fcmd, frc, fout, ferr]:
+            os.remove(f'/tmp/{f}')
+
+        res = {'returncode': int(contents[0]), 'stdout': contents[1], 'stderr': contents[2]}
+        log.debug(f'Real result: {res}')
+
+        return res
+
+    def _wrap_scp(self, src, dst):
+        return self._wrap_cmd(f'{SCP} {src} {dst}')
+
+    def put_file(self, src, dst=''):
+        cmd = self._wrap_scp(src, f'{self.user}@{self.ip}:{dst}')
+        return make_local_call(cmd, shell=True, timeout=self.timeout)
+
+    def get_file(self, src, dst='.'):
+        cmd = self._wrap_scp(f'{self.user}@{self.ip}:{src}', dst)
+        return make_local_call(cmd, shell=True, timeout=self.timeout)
+
+
+def ssh_command(ip, username, password, cmd_str, dbg_str=None, attempts=10, timeout=900):
+    """execute an SSH command, return both exit code, stdout and stderr."""
+    if dbg_str:
+        log.debug(dbg_str)
+
+    for i in range(0, attempts):
```

Comment:
such general retry can be a decorator,

I suppose this ssh retry does not ssh to the same server from lots of clients, otherwise, random interval needs to be used.

Reply:
Yes, this is just one to one ssh session.